### PR TITLE
feat: Convert dpl4hydra from shell script to C for performance

### PR DESCRIPTION
**Pull request by vschwaberow**
This commit significantly improves the functionality, performance, and user experience of the dpl4hydra tool, making it more efficient and easier to use for managing and generating default password lists.

### CODE REVIEW

**Comment by vschwaberow** [comment]
Build with:

`gcc -O3 -o dpl4hydra dpl4hydra.c -lcurl -lpthread`

File: dpl4hydra.c [hunk 1]
```diff
@@ -0,0 +1,503 @@
...
+  printf("  refresh     Download and refresh the full default password list\n");
+  printf("  list        List all available brands\n");
+  printf("  BRAND       Generate a default password list for a specific BRAND\n");
+  printf("  all         Generate a list of all system credentials\n\n");
+  printf("Example:\n");
+  printf("  %s linksys\n", program_name);
+}
+
+size_t write_callback(void *contents, size_t size, size_t nmemb, void *userp) {
+  size_t realsize = size * nmemb;
```
Comment (vanhauser-thc) [inline dpl4hydra.c:85]: unused var

File: dpl4hydra.c [hunk 2]
```diff
@@ -0,0 +1,503 @@
...
+
+  brands = malloc(capacity * sizeof(char *));
+  if (!brands) {
+    fprintf(stderr, "Error: Memory allocation failed.\n");
+    fclose(input);
+    exit(1);
+  }
+
+  while (fgets(line, sizeof(line), input)) {
+    if (sscanf(line, "%[^,]", brand) == 1) {
```
Comment (vanhauser-thc) [inline dpl4hydra.c:133]: buffer overflow

File: dpl4hydra.c [hunk 3]
```diff
@@ -0,0 +1,503 @@
...
+        if (!brands[brand_count]) {
+          fprintf(stderr, "Error: Memory allocation failed for brand.\n");
+          fclose(input);
+          for (int i = 0; i < brand_count; i++) {
+            free(brands[i]);
+          }
+          free(brands);
+          exit(1);
+        }
+        strcpy(prev_brand, brand);
```
Comment (vanhauser-thc) [inline dpl4hydra.c:159]: buffer overflow

**Comment by Logan0112** [comment]
Muito bom...parabéns

Em seg., 7 de out. de 2024 07:39, van Hauser ***@***.***>
escreveu:

**Comment by vschwaberow** [comment]
I'm putting the finishing touches on an update, coming your way soon. Is there anything else you'd like me to consider? I'm curious - do you find this feature valuable and worthwhile?

**Pull request closed without merging**